zerver/actions/custom_profile_fields.py

```diff
@@ -151,59 +157,66 @@ def do_update_user_custom_profile_data_if_changed(
+                )
+
+                # field_value.value is a TextField() so we need to have field["value"]
+                # in string form to correctly make comparisons and assignments.
+                if isinstance(custom_profile_field["value"], str):
+                    custom_profile_field_value_string = custom_profile_field["value"]
+                else:
+                    custom_profile_field_value_string = orjson.dumps(
+                        custom_profile_field["value"]
+                    ).decode()
+
+                if not created and field_value.value == custom_profile_field_value_string:
+                    # If the field value isn't actually being changed to a different one,
+                    # we have nothing to do here for this field.
+                    continue
+
+                field_value.value = custom_profile_field_value_string
+                if field_value.field.is_renderable():
+                    field_value.rendered_value = render_stream_description(
+                        custom_profile_field_value_string, user_profile.realm
+                    )
+                    field_value.save(update_fields=["value", "rendered_value"])
+                else:
+                    field_value.save(update_fields=["value"])
+                notify_user_update_custom_profile_data(
+                    user_profile,
+                    {
+                        "id": field_value.field_id,
+                        "value": field_value.value,
+                        "rendered_value": field_value.rendered_value,
+                        "type": field_value.field.field_type,
+                    },
                 )
-                field_value.save(update_fields=["value", "rendered_value"])
-            else:
-                field_value.save(update_fields=["value"])
-            notify_user_update_custom_profile_data(
-                user_profile,
-                {
-                    "id": field_value.field_id,
-                    "value": field_value.value,
-                    "rendered_value": field_value.rendered_value,
-                    "type": field_value.field.field_type,
-                },
-            )
 
 
 def check_remove_custom_profile_field_value(user_profile: UserProfile, field_id: int) -> None:
     try:
         custom_profile_field = CustomProfileField.objects.get(realm=user_profile.realm, id=field_id)
-        field_value = CustomProfileFieldValue.objects.get(
-            field=custom_profile_field, user_profile=user_profile
-        )
-        field_value.delete()
-        notify_user_update_custom_profile_data(
-            user_profile,
-            {
-                "id": field_id,
-                "value": None,
-                "rendered_value": None,
-                "type": custom_profile_field.field_type,
-            },
-        )
+        if not user_profile.is_realm_admin and not custom_profile_field.editable_by_user:
```

Comment:
These checks should be in the view function and should return error with appropriate message.